ddclient.in

```diff
@@ -2145,6 +2157,157 @@ sub geturl {
     return $reply;
 }
 
+######################################################################
+## geturl -- via curl
+######################################################################
+sub curl_cmd {
+    my @params = @_;
+    my $tmpfile;
+    my %curl_codes = (  ## Subset of error codes from https://curl.haxx.se/docs/manpage.html
+         3 => "URL malformed. The syntax was not correct",
+         5 => "Couldn't resolve proxy. The given proxy host could not be resolved.",
+         6 => "Couldn't resolve host. The given remote host was not resolved.",
+         7 => "Failed to connect to host.",
+        22 => "HTTP page not retrieved. The requested url was not found or returned another error.",
+        28 => "Operation timeout. The specified time-out period was reached according to the conditions.",
+        35 => "SSL connect error. The SSL handshaking failed.",
+        47 => "Too many redirects. When following redirects, curl hit the maximum amount.",
+        51 => "The peer's SSL certificate or SSH MD5 fingerprint was not OK.",
+        58 => "Problem with the local certificate.",
+        60 => "Peer certificate cannot be authenticated with known CA certificates.",
+        67 => "The user name, password, or similar was not accepted and curl failed to log in.",
+        77 => "Problem with reading the SSL CA cert (path? access rights?).",
+        78 => "The resource referenced in the URL does not exist.",
+    );
+
+    return '' if (scalar(@params) == 0); ## no parameters provided
+
+    # debug("CURL:\r\n%s\r\n",$params); ## Uncomment only if required as may expose credentials
+
+    eval "no warnings 'all'; use File::Temp;";
+    if (!$@) {
+        ## Use system File::Temp to generate filename...
+        $tmpfile = File::Temp->new(TMPDIR => 1,  ## in system temp dir
+                                   UNLINK => 0,  ## we will delete it ourselves
+                                 TEMPLATE => 'ddclient_XXXXXXXXXX');
+        ## File::Temp->new() automatically opens the file for us.
+    } else {
+        ## System does not have File::Temp so generate ourselves...
+        eval "no warnings 'all'; use File::Spec;";
+        $tmpfile = ((!$@) ? File::Spec->tmpdir() : '/tmp').'/ddclient_';
+        my @chars = ("A".."Z", "a".."z", "0".."9");
+        $tmpfile .= $chars[rand @chars] for 1..10;
+        open($tmpfile, '>', $tmpfile) or fatal("Falied to open $tmpfile for curl");
+    }
+    chmod(0600,$tmpfile);
+    debug("CURL Tempfile: %s",$tmpfile);
+    local $\ = "\n";  ## Terminate the file,
+    local $, = "\n";  ## and each parameter, with a newline.
+    print($tmpfile @params);
```

Comment:
Metacharacters in the entries of `@params` need to be quoted.

Reply:
My concern isn't with the location of the quoting, but with the lack of quoting. `fetch_via_curl` doesn't sanitize user input by quoting characters that are special in a curl config file (double quotes, newlines, spaces, etc.).

Reply:
> My concern isn't with the location of the quoting, but with the lack of quoting. `fetch_via_curl` doesn't sanitize user input by quoting characters that are special in a curl config file (double quotes, newlines, spaces, etc.).

I have added `quotemeta()` for the file path when not using `File::Temp` this should protect for when the system ENV is used.  I have also escaped those characters recognized by curl (see --config in curl [manpage](https://curl.haxx.se/docs/manpage.html))

Let me know if you see anything else that should be done. 

Thanks

Reply:
I think this is all resolved now.